Source/Particles/Filter/FilterFunctors.H

```diff
@@ -0,0 +1,120 @@
+/* Copyright 2016-2020 Maxence Thevenet, Yinjian Zhao
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#ifndef FILTERFUNCTORS_H
+#define FILTERFUNCTORS_H
+
+#include "WarpX.H"
+#include "Particles/WarpXParticleContainer.H"
+#include "Utils/WarpXConst.H"
+
+#include <AMReX_Gpu.H>
+#include <AMReX_Random.H>
+
+using SuperParticleType = typename WarpXParticleContainer::SuperParticleType;
+
+/**
+ * \brief Functor that returns 0 or 1 depending on a random draw per particle
+ */
+struct RandomFilter
+{
+    /** constructor
+     * \param a_is_active whether the test is active
+     * \param a_fraction fraction of particles to select
+     */
+    RandomFilter(bool a_is_active, amrex::Real a_fraction)
+        : m_is_active(a_is_active), m_fraction(a_fraction) {};
+
+    /**
+     * \brief draw random number, return 1 if number < m_fraction, 1 otherwise
+     * \param p one particle
+     * \return whether or not the particle is selected
+     */
+    AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+    bool operator () (const SuperParticleType& p) const noexcept
+    {
+        if ( !m_is_active ) return 1;
+        else if ( amrex::Random() < m_fraction ) return 1;
+        else return 0;
+    }
+private:
+    bool m_is_active = false; //! select all particles if false
+    amrex::Real m_fraction = 1.0; //! range: [0.0:1.0] where 0 is no & 1 is all particles
```

Comment:
As these are in the initializer list, they could be `const`.
```suggestion
    const bool m_is_active = false; //! select all particles if false
    const amrex::Real m_fraction = 1.0; //! range: [0.0:1.0] where 0 is no & 1 is all particles
```